Linear types amendment: Linear lets

aspiwack opened the pull request:
Would be best consumed using the rich diff view, except it doesn't appear to work. Fortunately, this is almost entirely added section, so I can point to them individually.

- [Specification of linear lets](https://github.com/tweag/ghc-proposals/blob/linear-lets/proposals/0111-linear-types.rst#let-and-where-bindings)
  - In addition to the above, `LinearTypes` is proposed to imply `MonoLocalBinds`
- [Effects and interaction](https://github.com/tweag/ghc-proposals/blob/linear-lets/proposals/0111-linear-types.rst#let-bindings-and-polymorphism)
- [Alternatives](https://github.com/tweag/ghc-proposals/blob/linear-lets/proposals/0111-linear-types.rst#id13)

This is a rather straightforward amendment which specifies multiplicity-annotated let bindings which, for reasons unknown to me, were only implied before. Pretty much all the choices are forced, the design space is really tiny. In consequence I intend to leave this open for comments a few work days before I submit it to the committee.

@simonpj , you may want to take part of the pre-committee reviews.

VitWW commented:
@aspiwack Maybe you forget to add **Rendered** link to your first comment

aspiwack commented:
@VitWW 

> Maybe you forget to add **Rendered** link to your first comment

I didn't: there isn't a rendered link because it's an amendment. Instead, go to [File changed](https://github.com/ghc-proposals/ghc-proposals/pull/624/files) and choose to display the rich diff. I'm afraid there isn't a better way.

PS: I had inadvertently included another proposal in this PR, I've fixed it as soon as it was pointed out to me.

aspiwack commented:
Except… the rich diff isn't working for me at the moment. Is it just me? Is it just my document?

googleson78 commented:
> Except… the rich diff isn't working for me at the moment. Is it just me? Is it just my document?

Doesn't work for me either at the moment

simonpj commented on proposals/0111-linear-types.rst:
```diff
@@ -1253,6 +1387,101 @@ interpret GADT-syntax type declaration (both ``data`` and ``newtype``)
 
+Let bindings and polymorphism
+-----------------------------
+
```
Refer back to the clause in the spec that says "linear bidnigns with multipiclity annotations are monomorphic"

And fwd ref to this section from the spec

JakobBruenker commented:
Can you have multiple bindings in one let? As in

```haskell
let %1 x = u
    %Many y = v
in …
```

aspiwack commented:
> Can you have multiple bindings in one let?

I don't see any reason not to. What makes you doubt it, so that I clarify in the proposal?

aspiwack commented:
I don't know what in my PR breaks the rich diff view, but I'm convinced at this point that it isn't transient. I've added link, in the PR description, to the various sections that I've added, hopefully I didn't forget any.

aspiwack commented:
The conversation died out. @nomeata I'd like to submit this proposal amendment to the committee.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+
+Non-variable linear patterns can't be lazy (see `Lazy patterns`_). As
+a consequence, non-variable let-bound patterns must be annotated with
+a ``!`` (because let-bound patterns are lazy by default, as opposed to
```
We still require a `!`. (we could do better, but it's unclear that it's worth it, especially since `-XStrict` is so rarely used)

goldfirere commented on proposals/0111-linear-types.rst:
```diff
@@ -2191,6 +2445,36 @@ would define
+are annotated with ``%Many``. But it invites a lot of complication
+(what if it's a type synonym that is equivalent to ``Many``? What if
+it's a type family?) that are hard to justify. In particular since we
+take ``-XMonoLocalBinds`` for granted, where generalisation doesn't
```
Again, this last point isn't accurate, but it doesn't matter. I'm fine with `let %Many` disabling generalization. If you want it, just use a type signature!

goldfirere reviewed (commented):
I've made a number of comments throughout. Many are simple typos or small corrections. Regardless, I'm in support of this -- especially because we are having debates about stability elsewhere, and thus seem to be carving out a space for more experimentation here. That is, maybe the interaction between generalization and the desire to infer linear bindings is a bit wrong in this proposal, something that will be hard to know until it's implemented. But that's fine: users of `-XLinearTypes` know that it's subject to change.

Small open question: how many users of `-XLinearTypes` do we have? If it's "lots", then this might want a new `-XLinearLets` extension that's less stable. But if it's "not that many, and we're friends with most of them", let's just keep it simple in one extension.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+   -- good
+   let %1 (Ur x) = u in (x, x, x)
+
+Non-variable let-bound patterns must be strict
```
This is covered by the section _Pattern-matching_ of the (unamended part of the) proposal. `~` patterns are unrestricted (maybe we should say “unrestricting” for a pattern as it's kind of a co-expression :slightly_smiling_face: ). View patterns are unrestricted as well, but not by design, rather by ignorance of what we could do better.

In this section, I only cover what's different for lets since patterns have a slightly different semantic in let-bindings from that in case-expressions.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+Non-variable linear patterns can't be lazy (see `Lazy patterns`_). As
+a consequence, non-variable let-bound patterns must be annotated with
+a ``!`` (because let-bound patterns are lazy by default, as opposed to
+case-bound patterns which are strict by default).
```
Sure. But we're talking about non-variable patterns here. Would it be clearer if I repeated it?

goldfirere replied:
Oh, right. OK.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+monomorphic types (see `Let bindings and polymorphism`_ for the
+reasoning and a discussion). Unannotated non-variable let-patterns are
+inferred to be unrestricted (by default, since ``-XLinearTypes``
+implies ``-XNoMonoLocalBinds``, only toplevel bindings, which are
```
Indeed, this was quite unfortunate. Thanks for catching it.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+reasoning and a discussion). Unannotated non-variable let-patterns are
+inferred to be unrestricted (by default, since ``-XLinearTypes``
+implies ``-XNoMonoLocalBinds``, only toplevel bindings, which are
+always unrestricted anyway, are inferred to be polymorphic).
```
I admit to having been a little careless in my phrasing here. Unless I'm missing something, this is all fine. Case (1) is the only one that does generalisation which, as you notice below, is the real problem. So in case 1 the binding needs to be unrestricted. But since it doesn't capture any variable, unrestricted is the most general type anyway. So we don't lose any generality.

Any suggestion for better phrasing?

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+
+::
+
+   {-# LANGUAGE NoMonoLocalBinds #-}
```
It makes a difference in the second to last example, actually.

Also, the fact that the first  two examples don't generalise despite `NoMonoLocalBinds` is part of what this meant to illustrate (though I guess that the way I wrote them, they would generalise even without `NoMonoLocalBinds`, weren't it for the modifier. How very frustrating: GHC is being too smart for my examples!). I've added some more explanation in the comments, plus a note about the special-casing of closed right-hand sides.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+
+   {-# LANGUAGE NoMonoLocalBinds #-}
+
+   -- x, y :: Int -> Int
```
Maybe? Does it make a difference?

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+   let %Many !(x, y) = ((\z->z), (\z->z)) in (x 0, x 1, y 2)
+
+   -- x :: forall a. a -> a
+   let %1 x = (\z -> z) in if b then x 1 else x Bool
```
Ah! But that's the very point! Linear _non-variable_ pattern bindings aren't generalised. Variable bindings, linear or not, can be freely generalised.

You've missed this point that wasn't meant to be particularly fine. It's worrying. Where do I insist on it to avoid confusion for future reader?

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+   -- x, y :: forall a. a -> a
+   let !(x, y) = ((\z->z), (\z->z)) in (x 0, x Bool, y 2)
+
+   -- rejected because x is inferred unrestricted but must be linear
```
Because it's a non-variable pattern binding and the type of variable is generalised. I added some more explanation in the comment. I very much assume that this will be surprising. Hence my adding the implication from LinearTypes to MonoLocalBinds.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -1160,6 +1315,8 @@ Linear wildcard patterns are disallowed.
 Lazy patterns
 ~~~~~~~~~~~~~
 
+.. _`Lazy patterns`
```
This is a _very_ interesting question. I'm not fully ready to answer the question yet. I'm doing some research on the subject these days.

What I can say is: I _believe_ that a convincing definition of semantic linearity (which doesn't quite exist yet) should tell you that yes, using `fst` once and `snd` once on a pair is linear (caveat: requires call-by-value or call-by-need. This isn't true in call-by-name semantics. Which should highlight that this isn't an obvious property). But making a linear type system (even an undecidable one) isn't quite obvious, if we understand type systems as being defined modularly over the structure of a term. Maybe it's possible. It's easy to allow linear lazy pattern matching in a type system, though, of course, and it should be fine (at least I believe it to be the case). But in the case of GHC, this isn't sufficient: we also need the desugaring to be linear, and that's where the difficulty lies.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -1253,6 +1410,103 @@ interpret GADT-syntax type declaration (both ``data`` and ``newtype``)
+
+For linearity the problem is that all these selections aren't
+linear and that ``p`` is called several times. It isn't because of lazy pattern
+matching (which is, independently, prohibited by the type-checker), a strict
```
I just forgot to add the `!`, I find that I'm still inconsistent at remembering them. I hope it eventually becomes a second nature…

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -1253,6 +1410,103 @@ interpret GADT-syntax type declaration (both ``data`` and ``newtype``)
+But this means pattern-matching on a lambda abstraction (albeit a lambda over
+type variables), which is not something that Core understands. Also, ``f`` and ``g``
+should be polymorphic, despite the fact that both fields of ``p`` are
+monomorphic. It's not clear that this makes sense. Even assuming that it makes
```
Right. The point was that the way we work around it for unrestricted bindings doesn't work for linear types. And that the way we compile linear pattern bindings doesn't apply to this situation.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -1253,6 +1410,103 @@ interpret GADT-syntax type declaration (both ``data`` and ``newtype``)
+prevents the type-checker from generating ``AbsBinds``, and, as such,
+makes more inferred lets linear, which is almost certainly the right default
+(at least it's the least surprising: a binding doesn't change from
+linear to unrestricted because a small change makes it generalisable).
```
I can't deny that there is this nagging part of my brain that suspects that maybe something is still amiss. But I haven't been able to find an issue yet.

aspiwack commented on proposals/0111-linear-types.rst:
```diff
@@ -2191,6 +2445,36 @@ would define
+patterns are monomorphic`_) that
+
+1. ``-XLinearTypes`` implies ``-XMonoLocalBinds``.
+2. Polymorphic non-multiplicity-annotated let-bound non-variable patterns are inferred to be unrestricted
```
I changed “Polymorphic” to “Generalised” hopefully this is clear.

aspiwack commented:
@goldfirere in addition to the threads above, the part about generalisation vs linearity was born of the implementation. I hadn't realised the tension prior to implementing all this.

goldfirere commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+if
+
+* The binding is not top-level
+* The binding is non-recursive
```
But I think

```
let %1 x = ... y ...
    y = ... x ...
```

is probably OK. But maybe this can't happen, because `x` cannot occur in the RHS of the unrestricted `y`. So perhaps I'm just making noise here. Resolve after reading.

goldfirere commented on proposals/0111-linear-types.rst:
```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
+proposal will be using the terminology “non-variable pattern”
+for patterns which aren't a single variable.
+
+Let bindings can optionally be annotated with a multiplicity
```
I like conservatism in this way in general, but I tend to think this is an easy place to accept a little more. I don't feel strongly.

aspiwack replied:
I don't feel very strongly either. I've added an alternative to that effect.

Because rendering is broken, let me copy in full what I wrote:

> Restrictions of multiplicity-annotated let bindings
> ---------------------------------------------------
> 
> The proposal specifies that a multiplicity annotated non-variable let binding ``let %p pat``
> must be such that ``pat = !pat'`` even if ``p = 'Many``. It is easy to
> lift this restriction on two dimension:
> 
> - We can say, instead, that patterns not of the form ``!pat'`` emit a
>   ``p ~ 'Many`` constraint instead. They already do (for the sake of
>   inference), so this is strictly less code.
> - We can generalise to more strict patterns. For instance, we don't
>   need to require a ``!`` if ``-XStrict`` is on, we can have patterns
>   of the form ``(!pat')`` (with additional parentheses). This is a few
>   lines of codes, inference actually already does this in my
>   implementation, so it's already paid for (though it does annoyingly
>   mostly duplicate another definition of strict pattern which I
>   couldn't find a way to factor as a single function, I don't like
>   this).
> 
> The reason that motivated the stronger restriction is to improve error
> messages, because we can then error out with “multiplicity-annotated
> let-bound patterns must be of the form !pat”, instead of the more
> mysterious “Couldn't unify 'Many with 'One”
> (see [#23586](https://gitlab.haskell.org/ghc/ghc/-/issues/23586)).
> But maybe the additional restrictions are more surprising than the
> error messages are helpful.
